### PR TITLE
feat: provide PoC on merging prediction outputs

### PR DESCRIPTION
**Pull request by Midnighter**
This provides a script to join one or more tables of AMP prediction. It's by no means complete and just a proof of concept based on work with @Darcy220606

### CODE REVIEW

File: modules/local/merge_predictions.nf [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    output = task.ext.prefix ?: 'predictions.tsv.gz'
+    """
+    merge_predictions.py \\
+        --output $output \\
+        --macrel $macrel \\
+        --ampir $ampir
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        python: \$(python --version | sed 's/Python //g')
```
Comment (jfy133) [inline modules/local/merge_predictions.nf:26]: It would be good to include versions of the packages as well (if possible)!

Comment (Midnighter) [reply modules/local/merge_predictions.nf:26]: Is that generally done in nf-core modules? You mean like pip list output?

Comment (jfy133) [reply modules/local/merge_predictions.nf:26]: I think I saw it somewhere.... Unless it was a local module... We can check another day. I might be wrong there

File: modules/local/merge_predictions.nf [hunk 2]
```diff
@@ -0,0 +1,29 @@
+process MERGE_PREDICTIONS {
+    tag "$samplesheet"
+
+    conda (params.enable_conda ? "conda-forge::python=3.8.6 conda-forge::pandas=1.0.5" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/mulled-v2-2076f4a3fb468a04063c9e6b7747a630abb457f6:fccb0c41a243c639e11dd1be7b74f563e624fcca-0' :
+        'quay.io/biocontainers/mulled-v2-2076f4a3fb468a04063c9e6b7747a630abb457f6:fccb0c41a243c639e11dd1be7b74f563e624fcca-0' }"
+
+    input:
+    tuple val(meta), path(macrel), path(ampir), ...
```
Comment (jfy133) [inline modules/local/merge_predictions.nf:10]:
```suggestion
    tuple val(meta), path(macrel), path(ampir)
```

File: subworkflows/local/amp_prediction.nf [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+include { MERGE_PREDICTIONS } from '../../modules/local/merge_predictions'
+
+workflow AMP_PREDICTION {
+    take:
+    proteins  // tuple val(meta), path('*.faa')
+
+    main:
+    MACREL_PEPTIDES(proteins)
+    AMPIR_PEPTIDES(proteins)
+    // AMPLIFY
```
Comment (jfy133) [inline subworkflows/local/amp_prediction.nf:16]:
```suggestion
```

Don't need for now

File: modules/local/merge_predictions.nf (hunk 1 above)
Comment (jfy133) [reply modules/local/merge_predictions.nf:26]: Ah it was a local module: https://github.com/nf-core/mag/blob/072b2f7be14376b4a943d3b1104c9c0e64038c41/modules/local/split_fasta.nf#L28-L32

So up to you whehter it's necessary

File: subworkflows/local/amp_prediction.nf [hunk 4]
```diff
@@ -0,0 +1,27 @@
+//
+// Check input samplesheet and get read channels
+//
+
+include { MACREL_CONTIGS } from '../../modules/nf-core/macrel/contigs/main'
+include { AMPIR_PEPTIDES } from '../../modules/foo'
```
Comment (jfy133) [inline subworkflows/local/amp_prediction.nf:6]:
```suggestion
include { AMPIR_PEPTIDES } from '../../modules/foo'
```

Remove debugging stuff in general if possible

File: subworkflows/local/amp_prediction.nf [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+include { MERGE_PREDICTIONS } from '../../modules/local/merge_predictions'
+
+workflow AMP_PREDICTION {
+    take:
+    proteins  // tuple val(meta), path('*.faa')
+
+    main:
+    MACREL_PEPTIDES(proteins)
```
Comment (jfy133) [inline subworkflows/local/amp_prediction.nf:14]: You load the module `MACREL_CONTIGS` not `MACREL_PEPTIDES?

File: subworkflows/local/amp_prediction.nf [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    AMPIR_PEPTIDES(proteins)
+    // AMPLIFY
+
+    ch_predictions = MACREL_CONTIGS.out.amp_prediction.join(
+        AMPIR_PEPTIDES.out.amp_prediction
+    )
+    // .join(...)
```
Comment (jfy133) [inline subworkflows/local/amp_prediction.nf:21]:
```suggestion
```

File: subworkflows/local/amp_prediction.nf [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+workflow AMP_PREDICTION {
+    take:
+    proteins  // tuple val(meta), path('*.faa')
+
+    main:
+    MACREL_PEPTIDES(proteins)
+    AMPIR_PEPTIDES(proteins)
+    // AMPLIFY
+
+    ch_predictions = MACREL_CONTIGS.out.amp_prediction.join(
```
Comment (jfy133) [inline subworkflows/local/amp_prediction.nf:18]: Will this work? How will you know which tool goes with with merge_predictions `--<tool>` flag?

**Review by jfy133 — CHANGES_REQUESTED** [review]
Script looks good AFAICS

However the nextflow code looks quite pseudo-codey, and I don't see yet how that would work.

I suspcet you will need to have one input element per tool, and insert the flag accordingly if there is a file present.

EDIT: reading OP is a good idea james :facepalm:

File: bin/merge_predictions.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:2]:
```suggestion
```

File: bin/merge_predictions.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+"""Provide a command line tool to merge AMP prediction tables."""
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:5]:
```suggestion
```

File: bin/merge_predictions.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+import argparse
+import logging
+from pathlib import Path
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:10]:
```suggestion
```

File: bin/merge_predictions.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+import pandas as pd
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:12]:
```suggestion
```

File: bin/merge_predictions.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    """
+    Join all tables together.
+
+    Tables are joined using an outer join which ensures that all sequence predictions
+    are in the resulting table but this may introduce missing values.
+
+    """
+    result = tables[0].join(tables[1:], how="outer", sort=True)
+    return result.reset_index()
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:27]:
```suggestion
```

File: bin/merge_predictions.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+def parse_macrel(filename):
+    """Parse tab-delimited macrel AMP predictions and return a homogenous table."""
+    df = pd.read_table(filename, sep="\t", skiprows=1)
+    df.rename(columns={"Access": "sequence_id", "AMP_probability": "macrel"})
+    return df[["sequence_id", "macrel"]].set_index("sequence_id")
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:34]:
```suggestion
```

File: bin/merge_predictions.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+def parse_ampir(filename):
+    """Parse tab-delimited ampir AMP predictions and return a homogenous table."""
+    # TODO: Define column names if the table doesn't have any.
+    df = pd.read_table(filename, sep="\t", header=None, names=[])
+    # df.rename(columns={"?": "sequence_id", "?": "ampir"})
+    return df[["sequence_id", "ampir"]].set_index("sequence_id")
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:42]:
```suggestion
```

File: bin/merge_predictions.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    )
+    parser.add_argument(
+        "-l",
+        "--log-level",
+        help="The desired log level (default WARNING).",
+        choices=("CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"),
+        default="WARNING",
+    )
+    return parser.parse_args(argv)
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:76]:
```suggestion
```

File: bin/merge_predictions.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+        tables.append(macrel_df)
+    if args.ampir:
+        assert args.ampir.is_file(), f"File not found: {args.ampir}"
+        logger.info("Parse ampir predictions.")
+        ampir_df = parse_ampir(args.ampir)
+        tables.append(ampir_df)
+    # TODO: add other prediction tools
+    result = join_probabilities(tables)
+    result.to_csv(args.output, header=True, index=True, sep="\t")
+
```
Comment (jasmezz) [inline bin/merge_predictions.py:98]:
```suggestion
```

**Comment by jfy133** [comment]
Superseded by the upcoming ampCOMBI

**Pull request closed without merging**